### PR TITLE
abstract most boiler-plate procedures into the library

### PR DESCRIPTION
**Pull request by bill-auger**
this version is refactored with as much of the boiler-plate as possible moved out of the front-end and into the library class

the front-end example.php is now simplified to the following flow:
  * instantiate the library object
  * check if previously authorized
    * if yes request data and present the returned data
    * if no request auth link and present that

in order to abstract storage away from the end user, token storage was moved to a new semi-abstract class which is intended to be subclassed for specifc data storage needs - one reference subclass exists for per-user $_SESSION storage and another for single-user flat text file storage

note that this version is incompatible with the previous version and so example.php was replaced with the re-written online-status.php example

### CODE REVIEW

**Comment by bill-auger** [comment]
also i have not tested it - it is probably broken :)

File: livecodingAuth.php [hunk 1]
```diff
@@ -8,46 +20,83 @@ class LivecodingAuth {
...
+      else if (empty($redirect_url)) throw new Exception(INVALID_REDIRECT_URL_MSG, 1);
+
       $this->client_id = $client_id;
       $this->client_secret = $client_secret;
       $this->state = uniqid();
-      $this->scope = $scope;
       $this->redirect_url = $redirect_url;
+      $this->scope=$scope;
+      $this->is_authorized=false;
+      $this->tokens = new LivecodingAuthTokens() ;
```
Comment (bill-auger) [inline livecodingAuth.php]: should be:
```php
$this->tokens = new LivecodingAuthTokensSession() ;
```

File: livecodingAuth.php [hunk 2]
```diff
@@ -8,46 +20,83 @@ class LivecodingAuth {
...
-    **/
-    public function setTokens($tokens) {
-      $this->access_token = $tokens->access_token;
-      $this->token_type = $tokens->token_type;
-      $this->expires_in = $tokens->expires_in;
-      $this->refresh_token = $tokens->refresh_token;
-      $this->scope = $tokens->scope;
+    public function fetchData($data_path) {
+      // Refresh tokens from API server if necessary
+      if ($this->tokens->is_stale()) $this->tokens->setTokens($this->refreshToken());
```
Comment (bill-auger) [inline livecodingAuth.php]: should be:
```php
if ($this->tokens->is_stale()) $this->refreshToken();
```
refreshToken() calls setTokens() implicitly and returns nil

File: livecodingAuth.php [hunk 3]
```diff
@@ -146,54 +193,138 @@ public function request($request) {
...
+    private function setExpiresIn($expires_in) {}
+
+    private function getScope() {}
+
+    private function setScope($scope) {}
+  }
+
+
+  class LivecodingAuthTokensSession extends LivecodingAuthTokens {
+
```
Comment (bill-auger) [inline livecodingAuth.php]: should call session_start() in constructor yes? in case the user does not

**Comment by Wapaca** [comment]
I created a development branch cause there are too much bugs. I corrected some of them but it stills not working properly.

**Comment by bill-auger** [comment]
a separate development branch is a good idea yes - except that now both branches are in exactly the same state - so if the development branch is broken (this is generally OK) then the master branch is also broken (this is generally not OK) - ideally the master branch would have remained in the working state where it was before this merge - these are the commands to put it back: 
```
git stash
git checkout master
git reset --hard 59c6be5b6156a11ad8ada309f98e3742153f1002
git push -f origin master
git checkout development
git stash pop
```
in case you were wondering - 59c6be5 is the commit ID of the last working commit on master (the one with log message "Removing $lctv_user in var_dump") - you can get these from running `git log` - after this the result will be the same as if you had left the master branch as it was, created the development branch first, and then merged the new commits only into the development branch 

generally it is not good practice to `push -f` on the master branch - this forcably re-writes the history which can lead to confusion if others had already downloaded the now clobbered version

**Comment by Wapaca** [comment]
Hi bill,

The current version is working in both branches. I fixed last bugs on development and tested it before merging to master.

**Comment by bill-auger** [comment]
ok great i was just about to try it out

**Pull request closed without merging**